### DISCUSSION TITLE
Matcher ideas

### DISCUSSION THREAD

**Discussion by EskiMojo14** [EskiMojo14/mix-n-matchers#3] [Ideas]
Thought it'd be nice to have a place for new matcher ideas. Feel free to chime in with ideas for matchers you've wanted :)

Current ideas:

- [x] `expect.arrayOnlyContaining([...vals])` - like `expect.arrayContaining`, but making sure that the received array is a subset of the expected array, rather than the other way around.
  - Implemented https://github.com/EskiMojo14/mix-n-matchers/pull/5
- [x] `expect.typeOf('function')` - like Vitest's `toBeTypeOf`, but asymmetric
  - Implemented https://github.com/EskiMojo14/mix-n-matchers/pull/4 
- [x] `expect.oneOf([1,2,3])`
  - Implemented https://github.com/EskiMojo14/mix-n-matchers/pull/7
- [x] `expect.enum(Direction)` and `expect(getDirection()).toBeEnum(Direction)`
  - Implemented https://github.com/EskiMojo14/mix-n-matchers/pull/8 
- [x] Sequence matchers for iterables
  - Implemented https://github.com/EskiMojo14/mix-n-matchers/pull/10
  - Extension idea - `toContainSequence(1,2,3)`, matching `[0,1,2,3]` or `[1,2,3,4]`

**Comment by aryaemami59** [comment]
How about assertions that could be further chained like `expect(someArray).toBeAnArrayOf('function').withLengthOf(2)`? A little complicated probably but it could be done.

**Reply by EskiMojo14** [reply]
Not possible due to how jest wraps matchers, I'm afraid - matchers always return a `{ pass: boolean, message: () => string }` shape and Jest decides whether to throw based on that. Nothing returned from the matcher gets exposed by the time you're calling `expect(value).toDoSomething()`.

**Comment by EskiMojo14** [comment]
`toHaveCallOrder`/`toHaveBeenCalledInOrder`: [twitter thread](https://x.com/kettanaito/status/1845794939197694105)

theoretically would work like `toEqualSequence` but for mock calls 

```ts
expect(mock).toHaveCallOrder([1], [2])
```
